install-rust-toolchain.sh

```diff
@@ -188,6 +191,26 @@ function install_esp_idf() {
     fi
 }
 
+function install_xtensa_elf_gcc() {
+    IDF_TOOL_XTENSA_ELF_GCC="${IDF_TOOLS_PATH}/tools/$1-gcc/${GCC_VERSION}-${ARCH}"
+    GCC_FILE="$1-gcc${GCC_VERSION}-${GCC_ARCH}.tar.gz"
+    GCC_DIST_URL="${GCC_DIST_MIRROR}/${GCC_PATCH}/${GCC_FILE}"
+    echo "* installing ${IDF_TOOL_XTENSA_ELF_GCC} "
+    if [ ! -d ${IDF_TOOL_XTENSA_ELF_GCC} ]; then
+        if [ ! -f "${GCC_FILE}" ]; then
+            echo "** Downloading ${GCC_DIST_URL}"
+            curl -LO "${GCC_DIST_URL}"
+        fi
+        mkdir -p "${IDF_TOOL_XTENSA_ELF_GCC}"
+        echo "IDF_TOOL_XTENSA_ELF_GCC ${IDF_TOOL_XTENSA_ELF_GCC}"
+        tar xf ${GCC_FILE} -C "${IDF_TOOL_XTENSA_ELF_GCC}" --strip-components=1
+        echo "done"
+    else
+        echo "already installed"
+    fi
+    IDF_TOOL_XTENSA_ELF_GCC_PATH="${IDF_TOOL_XTENSA_ELF_GCC}/bin/:${IDF_TOOL_XTENSA_ELF_GCC_PATH}"
```

Comment:
This will always leave an empty section at the end of the `$PATH` since the variable starts out empty.

```suggestion
    IDF_TOOL_XTENSA_ELF_GCC_PATH="${IDF_TOOL_XTENSA_ELF_GCC}/bin/${IDF_TOOL_XTENSA_ELF_GCC_PATH:+:${IDF_TOOL_XTENSA_ELF_GCC_PATH}}"
```

Reply:
You suggested change leads to `::` in some cases. I think that using `echo export PATH=\"${IDF_TOOL_XTENSA_ELF_GCC_PATH}\$PATH\"` in line [609](https://github.com/esp-rs/rust-build/blob/4d069cc9362d75017305608adc76fdb7e9b5b432/install-rust-toolchain.sh#L609) and [623](https://github.com/esp-rs/rust-build/blob/4d069cc9362d75017305608adc76fdb7e9b5b432/install-rust-toolchain.sh#L623) could be a solution

Reply:
The suggested change should never lead to `::`, as it's explicitly only appending a colon if the path is already non-empty.

Reply:
Or do you mean that it leads to `::` when sourcing the file in the case that `$PATH` was completely unset?  That's true, and I guess you could solve it that way, but I think it'd be cleaner to use a proper parameter expansion that conveys the intended behavior in both cases.

Reply:
Also, if we use: `install-rust-toolchain --build-target "esp32s2, esp32c3 esp32" ` we would have the following `$PATH` in `export-esp.sh:` 
```sh 
export PATH="/home/esp/.espressif/tools/xtensa-esp32-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin//home/esp/.espressif/tools/riscv32-esp-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin//home/esp/.espressif/tools/xtensa-esp32s2-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin/:+::+:/home/esp/.espressif/tools/xtensa-esp32s2-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin/:+::+:/home/esp/.espressif/tools/riscv32-esp-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin//home/esp/.espressif/tools/xtensa-esp32s2-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin/:+::+:/home/esp/.espressif/tools/xtensa-esp32s2-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin/:+::$PATH"

```

Reply:
The change as I suggested it would not lead to that result at all, did you accidentally only copy-past part of what I wrote?

```
$ git apply <<'EOF'
diff --git a/install-rust-toolchain.sh b/install-rust-toolchain.sh
index 7bd3b9e..709672d 100755
--- a/install-rust-toolchain.sh
+++ b/install-rust-toolchain.sh
@@ -208,7 +208,7 @@ function install_xtensa_elf_gcc() {
     else
         echo "already installed"
     fi
-    IDF_TOOL_XTENSA_ELF_GCC_PATH="${IDF_TOOL_XTENSA_ELF_GCC}/bin/:${IDF_TOOL_XTENSA_ELF_GCC_PATH}"
+    IDF_TOOL_XTENSA_ELF_GCC_PATH="${IDF_TOOL_XTENSA_ELF_GCC}/bin/${IDF_TOOL_XTENSA_ELF_GCC_PATH:+:${IDF_TOOL_XTENSA_ELF_GCC_PATH}}"
 }
 
 function install_rust_xtensa_toolchain() {
EOF

$ ./install-rust-toolchain.sh --build-target "esp32s2,esp32c3,esp32" --export-file export.sh

$ cat export.sh
```

```bash
export PATH="/home/kyrias/.espressif/tools/xtensa-esp32-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin/:/home/kyrias/.espressif/tools/riscv32-esp-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin/:/home/kyrias/.espressif/tools/xtensa-esp32s2-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-unknown-linux-gnu/bin/:$PATH"
export LIBCLANG_PATH="/home/kyrias/.espressif/tools/xtensa-esp32-elf-clang/esp-14.0.0-20220415-x86_64-unknown-linux-gnu/lib/"
```

Reply:
You are completely right, I don't know what was I doing wrong, but gave it another try today and it was working flawlessly, maybe I just needed some rest. Sorry for all the noise that I've caused and big thanks for your detailed explanations and help!

Reply:
It's no problem at all, our brains have a strong tendency to be very uncooperative.  :)